File: pw-app-kit/verify_onboarding.mjs
/**
 * verify_onboarding.mjs - PW App Kit v2 self-check.
 *
 * Run from the app folder after copying pw_access.js:
 *
 *   node verify_onboarding.mjs [path/to/.env] [optional_google_token]
 */

import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { join } from "node:path";
import { APP_NAME, PROXY_BASE_URL, checkAllowedStatus } from "./pw_access.js";

const envPath = process.argv[2] || ".env";
const token = process.argv[3] || "";
let ok = true;

const providerKeyHints = /GEMINI|MATHPIX|SARVAM|ELEVEN|ANTHROPIC|CLAUDE|LITELLM|OPENAI/i;
const oldPatternHints = ["log:false", '"log": false', "'log': false", "/api/usage-log", "UsageSession.flush", ".flush()"];
const skipDirs = new Set([".git", "node_modules", "__pycache__", "dist", "build", ".venv", "venv"]);

async function fetchWithTimeout(url, ms) {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), ms);
  try {
    return await fetch(url, { signal: ctrl.signal, headers: { connection: "close" } });
  } finally {
    clearTimeout(timer);
  }
}

function scanEnvForKeys(path) {
  const leaked = [];
  if (!existsSync(path)) return leaked;
  for (const line of readFileSync(path, "utf8").split(/\r?\n/)) {
    const s = line.trim();
    if (!s || s.startsWith("#") || !s.includes("=")) continue;
    const eq = s.indexOf("=");
    const name = s.slice(0, eq).trim();
    const value = s.slice(eq + 1).trim();
    if (value && providerKeyHints.test(name)) leaked.push(name);
  }
  return leaked;
}

function scanOldPatterns(root = ".") {
  const hits = [];
  function walk(dir) {
    for (const entry of readdirSync(dir)) {
      if (skipDirs.has(entry)) continue;
      const path = join(dir, entry);
      const st = statSync(path);
      if (st.isDirectory()) {
        walk(path);
        continue;
      }
      if (!/\.(py|js|jsx|ts|tsx|md)$/i.test(entry)) continue;
      const normalized = path.replaceAll("\\", "/");
      if (normalized.includes("pw-app-kit/")) continue;
      const text = readFileSync(path, "utf8");
      const found = oldPatternHints.find((hint) => text.includes(hint));
      if (found) hits.push(`${path}: ${found}`);
      if (hits.length >= 20) return;
    }
  }
  try {
    walk(root);
  } catch {
    return hits;
  }
  return hits;
}

console.log(`APP_NAME = ${JSON.stringify(APP_NAME)}`);
console.log(`PROXY    = ${PROXY_BASE_URL}`);

if (APP_NAME === "SET-YOUR-APP-NAME") {
  console.log("FAIL: APP_NAME is still the placeholder.");
  ok = false;
}

try {
  const r = await fetchWithTimeout(`${PROXY_BASE_URL}/api/apps`, 20000);
  const apps = r.ok ? (await r.json()).apps || [] : [];
  if (apps.includes(APP_NAME)) {
    console.log(`PASS: ${JSON.stringify(APP_NAME)} is registered on the proxy`);
  } else {
    console.log(`FAIL: ${JSON.stringify(APP_NAME)} not in /api/apps. Add its exact column to Whitelisted.`);
    ok = false;
  }
} catch (e) {
  console.log("WARN: could not reach /api/apps:", e.message);
  ok = false;
}

const leaked = scanEnvForKeys(envPath);
if (leaked.length) {
  console.log(`FAIL: provider keys still present in ${envPath}: ${leaked.join(", ")}`);
  ok = false;
} else {
  console.log(`PASS: no provider keys in ${envPath}`);
}

const oldHits = scanOldPatterns(".");
if (oldHits.length) {
  console.log("WARN: possible old combined-logging patterns found:");
  for (const hit of oldHits) console.log(`  ${hit}`);
  console.log("      Remove these unless they are intentional legacy/manual audit code.");
} else {
  console.log("PASS: no obvious old combined-logging patterns found");
}

if (token) {
  const status = await checkAllowedStatus(token);
  console.log(`allowlist check for supplied token: ${status}`);
  if (status === "error") console.log("  token invalid/expired, proxy unreachable, or server error");
  ok = ok && (status === "allowed" || status === "denied");
}

console.log("\nRESULT:", ok ? "ALL GOOD" : "ISSUES FOUND");
process.exitCode = ok ? 0 : 1;

File: pw-app-kit/verify_onboarding.py
"""
verify_onboarding.py - PW App Kit v2 self-check.

Run from the app folder after copying pw_access.py:

    python verify_onboarding.py [path/to/.env] [optional_google_token]

Checks:
  1. pw_access imports and APP_NAME is set.
  2. APP_NAME is registered on the proxy.
  3. No provider API keys remain in the given .env.
  4. Warns if common old combined-logging patterns appear nearby.
  5. Optional live allowlist check when a Google token is supplied.
"""

import os
import sys


PROVIDER_KEY_HINTS = (
    "GEMINI",
    "MATHPIX",
    "SARVAM",
    "ELEVEN",
    "ANTHROPIC",
    "CLAUDE",
    "LITELLM",
    "OPENAI",
)

OLD_PATTERN_HINTS = (
    "log:false",
    '"log": false',
    "'log': false",
    "/api/usage-log",
    "UsageSession.flush",
    ".flush()",
)


def scan_env_for_keys(env_path):
    leaked = []
    if not os.path.exists(env_path):
        return leaked
    with open(env_path, encoding="utf-8") as f:
        for line in f:
            s = line.strip()
            if not s or s.startswith("#") or "=" not in s:
                continue
            name, _, value = s.partition("=")
            if value.strip() and any(k in name.upper() for k in PROVIDER_KEY_HINTS):
                leaked.append(name.strip())
    return leaked


def scan_old_patterns(root="."):
    hits = []
    skip_dirs = {".git", "node_modules", "__pycache__", "dist", "build", ".venv", "venv"}
    for base, dirs, files in os.walk(root):
        dirs[:] = [d for d in dirs if d not in skip_dirs]
        for name in files:
            if not name.endswith((".py", ".js", ".jsx", ".ts", ".tsx", ".md")):
                continue
            path = os.path.join(base, name)
            try:
                text = open(path, encoding="utf-8", errors="ignore").read()
            except Exception:
                continue
            for hint in OLD_PATTERN_HINTS:
                if hint in text and "pw-app-kit" not in path.replace("\\", "/"):
                    hits.append(f"{path}: {hint}")
                    break
    return hits[:20]


def main():
    env_path = sys.argv[1] if len(sys.argv) > 1 else ".env"
    token = sys.argv[2] if len(sys.argv) > 2 else ""
    ok = True

    try:
        import pw_access
    except Exception as e:
        print("FAIL: cannot import pw_access:", e)
        sys.exit(1)

    print(f"APP_NAME = {pw_access.APP_NAME!r}")
    print(f"PROXY    = {pw_access.PROXY_BASE_URL}")

    if pw_access.APP_NAME == "SET-YOUR-APP-NAME":
        print("FAIL: APP_NAME is still the placeholder.")
        ok = False

    import requests

    try:
        r = requests.get(f"{pw_access.PROXY_BASE_URL}/api/apps", timeout=20)
        apps = r.json().get("apps", []) if r.status_code == 200 else []
        if pw_access.APP_NAME in apps:
            print(f"PASS: {pw_access.APP_NAME!r} is registered on the proxy")
        else:
            print(f"FAIL: {pw_access.APP_NAME!r} not in /api/apps. Add its exact column to Whitelisted.")
            ok = False
    except Exception as e:
        print("WARN: could not reach /api/apps:", e)
        ok = False

    leaked = scan_env_for_keys(env_path)
    if leaked:
        print(f"FAIL: provider keys still present in {env_path}: {', '.join(leaked)}")
        ok = False
    else:
        print(f"PASS: no provider keys in {env_path}")

    old_hits = scan_old_patterns(".")
    if old_hits:
        print("WARN: possible old combined-logging patterns found:")
        for hit in old_hits:
            print("  " + hit)
        print("      Remove these unless they are intentional legacy/manual audit code.")
    else:
        print("PASS: no obvious old combined-logging patterns found")

    if token:
        status = pw_access.check_allowed_status(token)
        print(f"allowlist check for supplied token: {status}")
        if status == "error":
            print("  token invalid/expired, proxy unreachable, or server error")
        ok = ok and status in ("allowed", "denied")

    print("\nRESULT:", "ALL GOOD" if ok else "ISSUES FOUND")
    sys.exit(0 if ok else 1)


if __name__ == "__main__":
    main()
